pkgutil.get_data support

`pkgutil.get_data` currently only returns `None` when called from a module py.test matches, which is inconvenient when trying to load test data from a file in a test package.

This happens because `AssertionRewritingHook` sets itself as the module's loader, but it does not implement the (optional) PEP302 `get_data` loader API. As a result, `pkgutil.get_data` simply returns `None` when invoked on a test module while running py.test.

It would be nice if either
* `AssertionRewritingHook` didn't set itself as the module's loader, though I don't know what the side-effects would be if it set `pkgutil.ImpLoader` as the loader
* or `AssertionRewritingHook` implemented `get_data`, for which the trivial implementation seems like it should work given the rewriting hook correctly sets `__file__` on the loaded module:

        def get_data(self, pathname):
            return open(pathname, "rb").read()

Yeah the problem is the test module being used as base to get the test data.

Create a `foo/__init__.py`, `foo/data.txt` with `Hey\n` and in `foo/test_foo.py`

```python
# contents of test_foo.py
import pkgutil
import unittest

class TestFoo(unittest.TestCase):
    def test_foo(self):
        assert pkgutil.get_data(self.__module__, 'data.txt') == 'Hey\n'
```
then from outside `foo`
```
> python -munittest foo.test_baz
.
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK

> py.test foo
============================= test session starts =============================
platform darwin -- Python 2.7.10 -- py-1.4.30 -- pytest-2.7.2
rootdir: test/foo, inifile: 
collected 1 items 

foo/test_baz.py F

================================ FAILURES ================================
________________________________ TestFoo.test_foo _______________________________

self = <foo.test_baz.TestFoo testMethod=test_foo>

    def test_foo(self):
>       assert pkgutil.get_data(self.__module__, 'data.txt') == 'Hey\n'
E       AssertionError: assert None == 'Hey\n'
E        +  where None = <function get_data at 0x10a9e1488>('foo.test_baz', 'data.txt')
E        +    where <function get_data at 0x10a9e1488> = pkgutil.get_data
E        +    and   'foo.test_baz' = <foo.test_baz.TestFoo testMethod=test_foo>.__module__

foo/test_baz.py:7: AssertionError
=========================== 1 failed in 0.03 seconds ===========================
```

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,6 +1,11 @@
 2.8.0.dev (compared to 2.7.X)
 -----------------------------
 
+- fix issue 808: pytest's internal assertion rewrite hook now implements the
+  optional PEP302 get_data API so tests can access data files next to them.
+  Thanks xmo-odoo for request and example and Bruno Oliveira for
+  the PR.
+
 - rootdir and inifile are now displayed during usage errors to help
   users diagnose problems such as unexpected ini files which add
   unknown options being picked up by pytest. Thanks to Pavel Savchenko for

--- a/_pytest/assertion/rewrite.py
+++ b/_pytest/assertion/rewrite.py
@@ -197,6 +197,12 @@ class AssertionRewritingHook(object):
         #  DefaultProvider is appropriate.
         pkg_resources.register_loader_type(cls, pkg_resources.DefaultProvider)
 
+    def get_data(self, pathname):
+        """Optional PEP302 get_data API.
+        """
+        with open(pathname, 'rb') as f:
+            return f.read()
+
 
 def _write_pyc(state, co, source_stat, pyc):
     # Technically, we don't have to have the same pyc format as

--- a/testing/test_assertrewrite.py
+++ b/testing/test_assertrewrite.py
@@ -664,6 +664,21 @@ class TestAssertionRewriteHookDetails(object):
             "* 1 passed*",
         ])
 
+    def test_get_data_support(self, testdir):
+        """Implement optional PEP302 api (#808).
+        """
+        path = testdir.mkpydir("foo")
+        path.join("test_foo.py").write(py.code.Source("""
+            class Test:
+                def test_foo(self):
+                    import pkgutil
+                    data = pkgutil.get_data('foo.test_foo', 'data.txt')
+                    assert data == b'Hey'
+        """))
+        path.join('data.txt').write('Hey')
+        result = testdir.runpytest()
+        result.stdout.fnmatch_lines('*1 passed*')
+
 
 def test_issue731(testdir):
     testdir.makepyfile("""

